File: JavaScript/5_DataStructures/10_Graph.js
/*

Two Way of Implementing the Graphs are using,
    => AdjacencyList   
    => AdjacentMatrix

    here we are gonna use Adjacency List
*/

class Graph {
  constructor() {
    this.AdjacencyList = {};
  }

  addVertex(vertex) {
    if (this.AdjacencyList[vertex]) return false;
    this.AdjacencyList[vertex] = [];
    return true;
  }

  removeVertex(vertex) {
    if (!this.AdjacencyList[vertex]) return false;
    this.AdjacencyList[vertex].forEach((connection) => {
      this.AdjacencyList[connection] = this.AdjacencyList[connection].filter(
        (connect_vertex) => vertex !== connect_vertex
      );
    });
    delete this.AdjacencyList[vertex];
    return true;
  }

  addEdge(vertex1, vertex2) {
    if (!(this.AdjacencyList[vertex1] || this.AdjacencyList[vertex2]))
      return false;
    this.AdjacencyList[vertex1].push(vertex2);
    this.AdjacencyList[vertex2].push(vertex1);
    return true;
  }

  removeEdge(vertex1, vertex2) {
    if (!(this.AdjacencyList[vertex1] || this.AdjacencyList[vertex2]))
      return false;
    this.AdjacencyList[vertex1] = this.AdjacencyList[vertex1].filter(
      (vertex) => vertex !== vertex2
    );
    this.AdjacencyList[vertex2] = this.AdjacencyList[vertex2].filter(
      (vertex) => vertex !== vertex1
    );
    return true;
  }

  depthFirstResursive(startVertex) {
    let result = [];
    let visited = {};

    let connections;

    const depthFirst = (start_vertex) => {
      connections = this.AdjacencyList[start_vertex];

      visited[start_vertex] = true;
      result.push(start_vertex);
      connections.forEach((vertex) => {
        if (!visited[vertex]) {
          depthFirst(vertex);
        }
      });
    };
    depthFirst(startVertex);

    return result;
  }

  depthfirstIterative(startVertex) {
    let result = [];
    let visited = {};
    let stack = [];
    stack.push(startVertex);
    visited[startVertex] = true;
    let vertex;

    while (stack.length) {
      vertex = stack.pop();

      result.push(vertex);
      this.AdjacencyList[vertex].forEach((connectionVertex) => {
        if (!visited[connectionVertex]) {
          stack.push(connectionVertex);
          visited[connectionVertex] = true; //visited should be marked once pushed to stack , or else it will append multiple times
        }
      });
    }

    return result;
  }

  breadthFirstIterative(startVertex) {
    let result = [];
    let visited = {};
    let queue = [];
    queue.push(startVertex);
    visited[startVertex] = true;
    let vertex;

    while (queue.length) {
      vertex = queue.shift();

      result.push(vertex);
      this.AdjacencyList[vertex].forEach((connectionVertex) => {
        if (!visited[connectionVertex]) {
          queue.push(connectionVertex);
          visited[connectionVertex] = true;
        }
      });
    }

    return result;
  }
}

let G = new Graph();
G.addVertex("A");
G.addVertex("B");
G.addVertex("C");
G.addVertex("D");
G.addVertex("E");
G.addVertex("F");

G.addEdge("A", "B");
G.addEdge("A", "C");
G.addEdge("B", "D");
G.addEdge("D", "E");
G.addEdge("D", "F");
G.addEdge("C", "E");
G.addEdge("E", "F");

/*
        A
    B       c
    D   -   E
        F


*/

File: Python/5_DataStructures/10_Graph.py
class Graph(object):
	def __init__(self):
		self.adjacencyList = {}

	def removeFromList(self, lst, val):
		for idx, _val in enumerate(lst):
			if val == _val:
				lst.pop(idx)
				break
		return lst

	def addVertex(self, vertex):
		if vertex in self.adjacencyList:
			return False

		self.adjacencyList[vertex] = []
		return True

	def removeVertex(self, vertex):
		if vertex not in self.adjacencyList:
			return False

		neighbours = self.adjacencyList[vertex]

		for neighbour in neighbours:
			self.adjacencyList[neighbour] = self.removeFromList(self.adjacencyList[neighbour], vertex)

		del self.adjacencyList[vertex]

		return True


	def addEdge(self, vertex1, vertex2):
		if vertex1 not in self.adjacencyList or vertex2 not in self.adjacencyList:
			return False

		if vertex2 in self.adjacencyList[vertex1]:
			return False

		self.adjacencyList[vertex1].append(vertex2)
		self.adjacencyList[vertex2].append(vertex1)
		return True
		

	def removeEdge(self, vertex1, vertex2):
		if vertex1 not in self.adjacencyList or vertex2 not in self.adjacencyList:
			return False

		self.adjacencyList[vertex1] = self.removeFromList(self.adjacencyList[vertex1], vertex2)
		self.adjacencyList[vertex2] = self.removeFromList(self.adjacencyList[vertex2], vertex1)

		return True

	def breadthFirst(self, vertex):
		if vertex not in self.adjacencyList: return False
		queue = [vertex]
		output = []
		visited = {}

		while len(queue) != 0:
			_vertex = queue.pop(0)
			if _vertex in visited: continue

			visited[_vertex] = True
			output.append(_vertex)
			queue += self.adjacencyList[_vertex]

		return output

	def breadthFirstRecursive(self, vertex):
		def recursiveHelper(queue = [vertex], output = [], visited = {}):
			if len(queue) == 0:return output

			_vertex = queue.pop(0)

			if _vertex not in visited:
				output.append(_vertex)

				visited[_vertex] = True
				queue += self.adjacencyList[_vertex]

			return recursiveHelper(queue, output, visited)
			
		return recursiveHelper([vertex])


	def depthFirst(self, vertex):
		def DFShelper(vertex, output = [], visited = {}):
			if vertex not in visited:
				output.append(vertex)
				visited[vertex] = True

				for _vertex in self.adjacencyList[vertex]:
					DFShelper(_vertex, output, visited)
			return output
		return DFShelper(vertex)

	def depthFirst2(self, vertex):
		output = []
		visited = {}
		def DFS(vertex):
			if vertex not in visited:
				output.append(vertex)
				visited[vertex] = True

				for _vertex in self.adjacencyList[vertex]:
					DFS(_vertex)
			return output
		return DFS(vertex)


		